pennylane/gradients/hadamard_gradient.py

```diff
@@ -309,218 +290,122 @@ def hadamard_grad(
+            generators_per_parameter.append(0)
             continue
 
-        trainable_op, idx, p_idx = tape.get_operation(trainable_param_idx)
+        trainable_op, idx, _ = tape.get_operation(trainable_param_idx)
 
         ops_to_trainable_op = tape.operations[: idx + 1]
         ops_after_trainable_op = tape.operations[idx + 1 :]
 
         # Get a generator and coefficients
-        sub_coeffs, generators = _get_generators(trainable_op)
+        sub_coeffs, generators = _get_pauli_generators(trainable_op)
         coeffs.extend(sub_coeffs)
-
-        num_tape = 0
+        generators_per_parameter.append(len(generators))
 
         for gen in generators:
-            if isinstance(trainable_op, qml.Rot):
-                # We only registered PauliZ as generator for Rot, therefore we need to apply some gates
-                # before and after the generator for the first two parameters.
-                if p_idx == 0:
-                    # Move the Rot gate past the generator
-                    op_before_trainable_op = ops_to_trainable_op.pop(-1)
-                    ops_after_trainable_op = [op_before_trainable_op] + ops_after_trainable_op
-                elif p_idx == 1:
-                    # Apply additional rotations that effectively move the generator to the middle of Rot
-                    ops_to_add_before = [
-                        qml.RZ(-trainable_op.data[2], wires=trainable_op.wires),
-                        qml.RX(np.pi / 2, wires=trainable_op.wires),
-                    ]
-                    ops_to_trainable_op.extend(ops_to_add_before)
-
-                    ops_to_add_after = [
-                        qml.RX(-np.pi / 2, wires=trainable_op.wires),
-                        qml.RZ(trainable_op.data[2], wires=trainable_op.wires),
-                    ]
-                    ops_after_trainable_op = ops_to_add_after + ops_after_trainable_op
-
             ctrl_gen = [qml.ctrl(gen, control=aux_wire)]
             hadamard = [qml.Hadamard(wires=aux_wire)]
             ops = ops_to_trainable_op + hadamard + ctrl_gen + hadamard + ops_after_trainable_op
 
-            measurements = []
-            # Add the Y measurement on the aux qubit
-            for m in tape.measurements:
-                if m.obs:
-                    obs_new = [m.obs]
-                else:
-                    m_wires = m.wires if len(m.wires) > 0 else tape.wires
-                    obs_new = [qml.Z(i) for i in m_wires]
-
-                obs_new.append(qml.Y(aux_wire))
-                obs_new = qml.prod(*obs_new)
-
-                if isinstance(m, qml.measurements.ExpectationMP):
-                    measurements.append(qml.expval(op=obs_new))
-                else:
-                    measurements.append(qml.probs(op=obs_new))
-
-            new_tape = qml.tape.QuantumScript(ops=ops, measurements=measurements, shots=tape.shots)
-            _rotations, _measurements = qml.tape.tape.rotations_and_diagonal_measurements(new_tape)
```

Comment:
We have zero reason to do this. We can handle observables in any pauli measurement basis just fine now.